### PR TITLE
Disable autocommit and manually commit filters in study view

### PR DESCRIPTION
**Pull request by BasLee**
Add a checkbox to the study page to disable 'auto submitting' of filters; and add a submit button that allows users to submit filters manually. ([Netlify preview](https://deploy-preview-4502--cbioportalfrontend.netlify.app/study/summary?id=blca_bcan_hcrn_2022))

image
_'Autosubmitting' of filters is enabled by default. It can be activated with the 'Manually submit' option, available in a new settings menu_

## Problem
When a user adds a new filter on the study page, it is automatically added and processed: data is requested from the backend and the panes are re-rendered. When studies grow and datasets get larger, study view panes become less responsive, making it quite laborious to add a few filters. 

## Solution
This PR adds a checkbox and a submit button. The checkbox allows users to disable 'autosubmitting'. New filters are now queued. Once the new submit button is clicked, queued filters are submitted all at once, only triggering a single rerender.

### Disabling autocommit

image

When disabling autosubmit, filters are added to the queue. Panes are only updated when the submit button is clicked.

### Visualizing submitted and queued filters

image

When disabling autosubmit, the filter bar color codes 1) newly added filters; 2) already applied filters; and 3) filters that are queued for deletion.

## TODO
- [x] Move toggle to gear menu
- [x] Show which filters are submitted and which are not 
- [x] Add e2e test

### CODE REVIEW

**Comment by BasLee** [comment]
summary of feedback during refactorers stand-up:
- move auto submit toggle to new gear button
  - more study view options can be moved there later on
- move submit button closer to filters, before 'clear all filters' button
- visualize which filters are applied and which are not yet submitted
  - add new line with queued filters (show submit button on this line?)

**Comment by pvannierop** [comment]
[Disable auto-commit and manually commit filters in Study View](https://trello.com/c/dQhc8jTI/7-disable-auto-commit-and-manually-commit-filters-in-study-view)

File: src/shared/components/PillTag/PillTag.tsx [hunk 1]
```diff
@@ -3,19 +3,67 @@ import _ from 'lodash';
...
+ * also contains an onDeleteCallback, which will be executed on submit.
+ *
+ * Filters can be of different shapes, and they are created
+ * in a number of different places, so we use these two global stores
+ * to keep track of hesitant (queued) filters and submitted filters.
+ *
+ * TODO: A better solution would be a uniform list of filters
+ *  but this would require some refactoring.
+ */
+getBrowserWindow().hesitantPillStore = {} as PillStore;
```
Comment (alisman) [inline src/shared/components/PillTag/PillTag.tsx:48]: why not put these in the StudyView context?

Comment (alisman) [reply src/shared/components/PillTag/PillTag.tsx:48]: also, is "hesitant" a typo?

**Comment by alisman** [comment]
@pvannierop @BasLee i did some testing here.  I think there are probably some more tweaks that we would want to make to the UI before going live with this for our own users.
- The "pending" style is not obvious enough
- The hesitate mode state should be persisted?
- The product team may want to weigh in on the placement of the submit button
- Right now when you toggle the hesitate mode, some of the charts re-render, which is odd. We should see if we can prevent that.

I think a feature like this should be manually tested thoroughly by someone on your team who is not a developer.  

It should also be well covered by e2e tests.

File: src/shared/components/PillTag/PillTag.tsx (hunk 1 above)
Comment (BasLee) [reply src/shared/components/PillTag/PillTag.tsx]: Fixed in [60e31e2](https://github.com/cBioPortal/cbioportal-frontend/pull/4502/commits/60e31e2d279ad9e2f6e6a0ff18bac2d4199c64f6) (pill stores are now in study view store, and study view store is passed to pills using the context)

**Comment by BasLee** [comment]
@alisman https://github.com/cBioPortal/cbioportal-frontend/pull/4502#issuecomment-1450463645
Added local e2e tests in  [6e33cd4](https://github.com/cBioPortal/cbioportal-frontend/pull/4502/commits/6e33cd4ced8f09b697da179f36b71f7149cd975a)

File: src/pages/studyView/StudyViewPageStore.ts [hunk 2]
```diff
@@ -466,6 +480,44 @@ export class StudyViewPageStore
             )
         );
 
+        this.reactionDisposers.push(
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:483]: i suspect that it is the reactions which may be causing the charts to re-render whenever hesitateUpdate changes.

File: src/pages/studyView/StudyViewPageStore.ts [hunk 3]
```diff
@@ -466,6 +480,44 @@ export class StudyViewPageStore
...
+                        const onDeleteCallback = value.onDeleteCallback;
+                        if (onDeleteCallback) {
+                            onDeleteCallback();
+                            delete this.submittedPillStore[key];
+                        } else {
+                            this.submittedPillStore[key] = value;
+                        }
+                    });
+                    this.hesitantPillStore = {};
+                    this.filters = this.filtersProx;
```
Comment (alisman) [inline src/pages/studyView/StudyViewPageStore.ts:516]: maybe it's this here?  this is going to update filters every time hesitateUpdate is changed, even when filters are exactly the same.  just a hunch.

**Comment by BasLee** [comment]
image

☝️Feedback from community call on March 7

**Pull request merged**